src/misc/2fa/webauthn/WebauthnClient.ts

```diff
@@ -1,167 +1,149 @@
-		}
-		let assertion
-
-		try {
-			assertion = await this.api.get({
-				publicKey: publicKeyCredentialRequestOptions,
-				signal,
-			})
-		} catch (e) {
-			if (e.name === "AbortError") {
-				throw new WebauthnCancelledError(e)
+			keys: allowedKeys,
+			domain: this.selectAuthenticationUrl(challenge)
+		})
+
+		return createWebauthnResponseData({
+			keyHandle: new Uint8Array(signResult.rawId),
+			clientData: new Uint8Array(signResult.clientDataJSON),
+			signature: new Uint8Array(signResult.signature),
+			authenticatorData: new Uint8Array(signResult.authenticatorData),
+		})
+	}
+
+	abortCurrentOperation(): Promise<void> {
+		return this.webauthn.abortCurrentOperation()
+	}
+
+	private selectAuthenticationUrl(challenge: U2fChallenge): string {
+		// We need to figure our for which page we need to open authentication based on the keys that user has added because users can register keys for our
+		// domains as well as for whitelabel domains.
+
+		let selectedClientUrl
+		if (challenge.keys.some(k => k.appId === WEBAUTHN_RP_ID)) {
+			// First, if we find our own key then open web client on our URL.
+			// Even if it's a different subdomain of ours it can still match because it is scoped for all tutanota.com subdomains
+			selectedClientUrl = this.clientWebRoot
+		} else {
+			// If it isn't there, look for any Webauthn key. Legacy U2F key ids ends with json subpath.
+			const webauthnKey = challenge.keys.find(k => !this.isLegacyU2fKey(k))
+			if (webauthnKey) {
+				selectedClientUrl = `https://${webauthnKey.appId}`
+			} else if (challenge.keys.some(k => k.appId === U2F_APPID)) {
+				// There are only legacy U2F keys but there is one for our domain, take it
+				selectedClientUrl = this.clientWebRoot
 			} else {
-				throw new WebauthnError(e)
+				// Nothing else worked, select legacy U2F key for whitelabel domain
+				selectedClientUrl = this.legacyU2fKeyToBaseUrl(firstThrow(challenge.keys))
 			}
 		}
+		return selectedClientUrl
+	}
 
-		const publicKeyCredential: PublicKeyCredential | null = downcast(assertion)
+	private isLegacyU2fKey(key: U2fKey): boolean {
+		return key.appId.endsWith(U2f_APPID_SUFFIX)
+	}
 
-		if (publicKeyCredential == null) {
-			throw new ProgrammingError("Webauthn credential could not be unambiguously resolved")
-		}
+	private legacyU2fKeyToBaseUrl(key: U2fKey): string {
+		assert(this.isLegacyU2fKey(key), "Is not a legacy u2f key")
 
-		const response: AuthenticatorAssertionResponse = downcast(publicKeyCredential.response)
-		return createWebauthnResponseData({
-			keyHandle: new Uint8Array(publicKeyCredential.rawId),
-			clientData: new Uint8Array(response.clientDataJSON),
-			signature: new Uint8Array(response.signature),
-			authenticatorData: new Uint8Array(response.authenticatorData),
-		})
+		return key.appId.slice(0, -(U2f_APPID_SUFFIX.length))
 	}
 
-	_getChallenge(): Uint8Array {
+	private getChallenge(): Uint8Array {
 		// Should be replaced with our own entropy generator in the future.
```

Comment:
Does in the future mean "before we release this" or just "at our leisure"? should we attend to it?

Reply:
"Eventually". Sounds like a Friday project. It is left over initial debating review